lib/rails_twirp/testing/integration_test.rb

```diff
@@ -17,25 +23,33 @@ def before_rpc(&block)
       @before_rpc << block
     end
 
-    def rpc(service, rpc, request, headers: nil)
+    def rpc(svc, rpc, request, headers: nil)
       @request = request
-      service = app.twirp.routes.services[service].to_service
+      input_class = svc.rpcs[rpc][:input_class]
+      output_class = svc.rpcs[rpc][:output_class]
 
-      rack_env = {
-        "HTTP_HOST" => "localhost"
+      content_type = Twirp::Encoding::PROTO
+      env = {
+        "CONTENT_TYPE" => content_type,
+        "HTTP_HOST" => "localhost",
+        "PATH_INFO" => "#{@mount_path}/#{svc.service_full_name}/#{rpc}",
+        "REQUEST_METHOD" => "POST"
       }
-      http_request = ActionDispatch::Request.new(rack_env)
+      http_request = ActionDispatch::Request.new(env)
       http_request.headers.merge! headers if headers.present?
-      env = {rack_env: rack_env}
+
+      env["rack.input"] = StringIO.new(Twirp::Encoding.encode(request, input_class, content_type))
```

Comment:
As discussed offline, could you extract the building of the `twirp_request` into a helper?